source/extensions/clusters/redis/redis_cluster.h

```diff
@@ -0,0 +1,231 @@
+                             parent.lbEndpoint().endpoint().health_check_config(),
+                             parent.localityLbEndpoint().priority(),
+                             parent.lbEndpoint().health_status()),
+          master_(master) {}
+
+    bool isMaster() const { return master_; }
+
+  private:
+    const bool master_;
+  };
+
+  // Resolve the discovery endpoint
+  struct DnsDiscoveryResolveTarget {
+    DnsDiscoveryResolveTarget(
+        RedisCluster& parent, const std::string& dns_address,
+        const envoy::api::v2::endpoint::LocalityLbEndpoints& locality_lb_endpoint,
+        const envoy::api::v2::endpoint::LbEndpoint& lb_endpoint);
+
+    ~DnsDiscoveryResolveTarget();
+
+    void startResolve();
+
+    RedisCluster& parent_;
+    Network::ActiveDnsQuery* active_query_{};
+    std::string dns_address_;
+    const envoy::api::v2::endpoint::LocalityLbEndpoints locality_lb_endpoint_;
+    const envoy::api::v2::endpoint::LbEndpoint lb_endpoint_;
+  };
+
+  typedef std::unique_ptr<DnsDiscoveryResolveTarget> DnsDiscoveryResolveTargetPtr;
+
+  struct RedisDiscoverySession
+      : public Extensions::NetworkFilters::Common::Redis::Client::Config,
+        public Extensions::NetworkFilters::Common::Redis::Client::PoolCallbacks,
+        public Network::ConnectionCallbacks {
+    RedisDiscoverySession(RedisCluster& parent,
+                          NetworkFilters::Common::Redis::Client::ClientFactory& client_factory);
+
+    ~RedisDiscoverySession();
+
+    void registerDiscoveryAddress(
+        const std::list<Network::Address::InstanceConstSharedPtr>& address_list);
+
+    // Start discovery against a random host from existing hosts
+    void startResolve();
+
+    // Extensions::NetworkFilters::Common::Redis::Client::Config
+    bool disableOutlierEvents() const override { return true; }
+    std::chrono::milliseconds opTimeout() const override {
+      // Allow the main Health Check infra to control timeout.
+      return parent_.cluster_refresh_timeout_;
+    }
+    bool enableHashtagging() const override { return false; }
+    bool enableRedirection() const override { return false; }
+
+    // Extensions::NetworkFilters::Common::Redis::Client::PoolCallbacks
+    void onResponse(NetworkFilters::Common::Redis::RespValuePtr&& value) override;
+    void onFailure() override;
+    // Note: Below callback isn't used in topology updates
+    bool onRedirection(const NetworkFilters::Common::Redis::RespValue&) override { return true; }
+
+    // Network::ConnectionCallbacks
+    void onEvent(Network::ConnectionEvent event) override;
+    void onAboveWriteBufferHighWatermark() override {}
+    void onBelowWriteBufferLowWatermark() override {}
+
+    RedisCluster& parent_;
+    Extensions::NetworkFilters::Common::Redis::Client::ClientPtr client_;
+    Extensions::NetworkFilters::Common::Redis::Client::PoolRequest* current_request_{};
+    absl::Mutex discovery_mutex_;
```

Comment:
Can you describe why the mutex is necessary? In general all topology changes should be on the main thread and we shouldn't need any mutexes here.

Reply:
That's a good point- I'll remove that.